api/app/src/main/kotlin/packit/service/JwtLoginService.kt

```diff
@@ -0,0 +1,103 @@
+package packit.service
+
+import org.slf4j.LoggerFactory
+import org.springframework.http.HttpStatus
+import org.springframework.security.oauth2.core.DelegatingOAuth2TokenValidator
+import org.springframework.security.oauth2.core.oidc.IdTokenClaimNames
+import org.springframework.security.oauth2.jwt.Jwt
+import org.springframework.security.oauth2.jwt.JwtClaimValidator
+import org.springframework.security.oauth2.jwt.JwtException
+import org.springframework.security.oauth2.jwt.JwtIssuerValidator
+import org.springframework.security.oauth2.jwt.JwtTimestampValidator
+import org.springframework.security.oauth2.jwt.NimbusJwtDecoder
+import org.springframework.stereotype.Component
+import org.springframework.web.client.RestOperations
+import org.springframework.web.client.RestTemplate
+import packit.config.JwtLoginConfig
+import packit.config.JwtPolicy
+import packit.exceptions.PackitException
+import packit.model.dto.LoginWithToken
+import packit.security.provider.JwtIssuer
+import java.util.function.Predicate
+
+data class TokenPolicy(
+  val decoder: NimbusJwtDecoder,
+  val config: JwtPolicy,
+)
+
+@Component
+class JwtLoginService(
+  val jwtIssuer: JwtIssuer,
+  val userService: UserService,
+  val jwtLoginConfig: JwtLoginConfig,
+  val restOperations: RestOperations = RestTemplate(),
+) {
+  lateinit var policies: List<TokenPolicy>
+
+  val audience: String? = jwtLoginConfig.audience
+
+  init {
```

Comment:
all the init should probs be in `JwtLoginConfig` and then it has the policies it can pass.. and thus we can keep the service clean

Reply:
I'm not sure I agree. The `JwtLoginConfig` is a plain data class that just represents the properties. It seems odd that it would be initializing some business logic by creating validators and JWT decoders.

Reply:
yeah that is fair but it seems like it does derive directly from config and doesn't need anything else right? and yeah the init in the service seems a bit odd.. maybe could create a bean with this logic and then directly inject into jwtlogin service?

something like this maybe? 
```
@Configuration
class JwtConfig {
    @Bean
    fun tokenPolicies(jwtLoginConfig: JwtLoginConfig, restOperations: RestOperations): List<TokenPolicy> {
        // init logic
    }
}
```

then can pass into service straight up
```
class JwtLoginService(
    val policies: List<TokenPolicy>,
 val jwtIssuer: JwtIssuer,
    val userService: UserService,
)
```


but if ya disagree feel free to keep your way :smile:

Reply:
I think I'll keep it this way. The contents of the `TokenPolicy` is so tightly coupled with the implementation of the service it doesn't really make sense to move it elsewhere. The only way the decoupling makes sense IMO is if we make the TokenPolicy an interface and move all the logic into it, at which point the service is nothing much more than a for-loop. Seems like an unnecessary abstraction, especially given there's only one kind of TokenPolicy.

I've changed it from a `lateinit var` to a `val` though, since I realised that works just as well.